### PR TITLE
Pods 7543

### PR DESCRIPTION
**Pull request by grantrobson**

### CODE REVIEW

File: test/controllers/address/EnterPostcodeControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+      val application =
+        applicationBuilder(userAnswers = Some(emptyUserAnswers), extraModules)
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, postRoute).withFormUrlEncodedBody(("value", ""))
+
+        val view = application.injector.instanceOf[EnterPostcodeView]
```
Comment (cyeniceri) [inline test/controllers/address/EnterPostcodeControllerSpec.scala:120]: view and boundForm are not compared and waiting as unused here.

File: test/controllers/address/ChooseAddressControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+      val ua = emptyUserAnswers.setOrException(EnterPostcodePage(Event1EmployerAddressJourney), seqTolerantAddresses)
+
+      val application =
+        applicationBuilder(userAnswers = Some(ua), extraModules)
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, postRoute).withFormUrlEncodedBody(("value", "invalid"))
```
Comment (cyeniceri) [inline test/controllers/address/ChooseAddressControllerSpec.scala:127]: shouldn't invalid-value be "" for these radio-button list ? and we may validate form and result ?

**Review by cyeniceri — CHANGES_REQUESTED** [review]

File: app/models/address/Address.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+case class Address(addressLine1: String,
+                   addressLine2: String,
+                   addressLine3: Option[String],
+                   addressLine4: Option[String],
+                   postcode: Option[String],
+                   country: String) {
+
+  def print: String = {
```
[the remark this answers is not in the document]
Comment (grantrobson) [reply app/models/address/Address.scala:35]: Probably not, I just copied from scheme. I will remove

File: app/models/address/TolerantAddress.scala [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    Address(
+      addressLine1.getOrElse(""),
+      addressLine2.getOrElse(""),
+      addressLine3,
+      addressLine4,
+      postcode,
+      countryOpt.getOrElse("")
+    )
+
+  def toPrepopAddress: Address = toAddress.getOrElse(prepopAddress)
```
[the remark this answers is not in the document]
Comment (grantrobson) [reply app/models/address/TolerantAddress.scala:54]: I will remove

File: app/models/enumeration/AddressJourneyType.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    eventType = EventType.Event1,
+    nodeName = "employerAddress",
+    entityTypeMessageKey = entityTypeMessageKeyCompany) {
+    override def entityName(ua: UserAnswers): Message = ua.get(CompanyDetailsPage) match {
+      case Some(cd) => Literal(cd.companyName)
+      case _ => Message(entityTypeMessageKeyCompany)
+    }
+  }
+
+  // TODO: Remove this dummy object when we have at least two usages of the AddressJourneyType. If only one instance then we get compile errors.
```
[the remark this answers is not in the document]
Comment (grantrobson) [reply app/models/enumeration/AddressJourneyType.scala:60]: Just means two instances of the class. I have updated comment

File: test/controllers/address/ChooseAddressControllerSpec.scala (hunk 2 above)
Comment (grantrobson) [reply test/controllers/address/ChooseAddressControllerSpec.scala:124]: Either would be invalid. I will change to empty string

File: app/views/address/ManualAddressView.scala.html [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            )
+            .withWidth(OneQuarter)
+        )
+
+        <div class="govuk-!-width-one-half">
+            @govukSelect(Select(
+                id = "country",
+                name = "country",
+                classes = "govuk-!-width-one-half country",
+                items = Seq(SelectItem(value = Some(""))) ++
```
[the remark this answers is not in the document]
Comment (grantrobson) [reply app/views/address/ManualAddressView.scala.html:91]: Probably this behaviour will be the same in other services.

File: test/controllers/address/ChooseAddressControllerSpec.scala (hunk 2 above)
Comment (grantrobson) [reply test/controllers/address/ChooseAddressControllerSpec.scala]: We don't really need to check result. So long as it is a BAD REQUEST we know it has failed validation. The form spec will test the specific error.

**Review by cyeniceri — APPROVED** [review]

**Pull request merged**